In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats

# Question:

In a company there are 3 shifts i.e shift1, shift2, shift3 & we have 3 operators & we had production number of these 3 operators for these 3 shifts. Our question is, Is there any relationship between operator & shift? Test with 95% confidence level.

<table>
    <tr>
        <th>  </th>
        <th> Operation 1 </th>
        <th> Operation 2 </th>
        <th> Operation 3 </th>
        <th>  </th>
    </tr>
     <tr>
        <td> <b>Shift1</b> </td>
        <td> 22 </td>
        <td> 26 </td>
        <td> 23 </td>
        <td> <b>71</b> </td>
<!--         <td> 4.3 </td> -->
    </tr>
    <tr>
        <td> <b>Shift2</b> </td>
        <td> 28 </td>
        <td> 62 </td>
        <td> 26 </td>
        <td> <b>116</b> </td>
<!--         <td> 4 </td> -->
    </tr>
    <tr>
        <td> <b>Shift3</b> </td>
        <td> 72 </td>
        <td> 22 </td>
        <td> 66 </td>
        <td> <b>160</b> </td>
<!--         <td> 4.5 </td> -->
    </tr>
    <tr>
        <td>  </td>
        <td> <b>122</b> </td>
        <td> <b>110</b> </td>
        <td> <b>115</b> </td>
        <td> <b>347</b> </td>
<!--         <td> 5.1 </td> -->
    </tr>
</table>


In [2]:
sh_op = np.array([[22, 26, 23], [28, 62, 26], [72, 22, 66]])
sh_op

array([[22, 26, 23],
       [28, 62, 26],
       [72, 22, 66]])

In [3]:
stats.chi2_contingency(sh_op)

(50.09315721064659,
 3.4527076339398545e-10,
 4,
 array([[24.96253602, 22.50720461, 23.53025937],
        [40.78386167, 36.77233429, 38.44380403],
        [56.25360231, 50.7204611 , 53.0259366 ]]))

- Chi_Square Calculated: 50.09315721064659

- P-Value: 3.4527076339398545e-10

- DegreeOfFreedom: 4

Now Compare P-Value with Alpha i.e 0.05.

- We notice that P-Value < Alpha, Therefore **REJECT** H0.

            (YES, there a relationship between operator & shift)

<br>
_____________________________________________________________________________________________________________________________

## Same Question But Using DataFrame Bcz In Real Life DataFrame Will Be Given

In [4]:
sh_op_df = pd.DataFrame(sh_op, columns=['Shift1', 'Shift2', 'Shift3'], index=['Worker1', 'Worker2', 'Worker3'])
sh_op_df

Shift1  Shift2  Shift3
Worker1      22      26      23
Worker2      28      62      26
Worker3      72      22      66

In [5]:
stats.chi2_contingency(sh_op_df)

(50.09315721064659,
 3.4527076339398545e-10,
 4,
 array([[24.96253602, 22.50720461, 23.53025937],
        [40.78386167, 36.77233429, 38.44380403],
        [56.25360231, 50.7204611 , 53.0259366 ]]))

**Same Result As Above**

Now Compare P-Value with Alpha i.e 0.05.

- We notice that P-Value < Alpha, Therefore **REJECT** H0.

            (YES, there a relationship between operator & shift)

<br>
_____________________________________________________________________________________________________________________________

# Question 2: 

Is there is a relationship between Days and the Smokers?

In [6]:
tips = sns.load_dataset('tips')
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [7]:
tips.groupby(by = ['day', 'smoker']).count()

total_bill  tip  sex  time  size
day  smoker                                  
Thur Yes             17   17   17    17    17
     No              45   45   45    45    45
Fri  Yes             15   15   15    15    15
     No               4    4    4     4     4
Sat  Yes             42   42   42    42    42
     No              45   45   45    45    45
Sun  Yes             19   19   19    19    19
     No              57   57   57    57    57

In [8]:
day_smoker = tips.pivot_table(index='day', columns='smoker', aggfunc='count')['tip']
day_smoker

smoker  Yes  No
day            
Thur     17  45
Fri      15   4
Sat      42  45
Sun      19  57

In [9]:
stats.chi2_contingency(day_smoker)

(25.787216672396262,
 1.0567572499836523e-05,
 3,
 array([[23.63114754, 38.36885246],
        [ 7.24180328, 11.75819672],
        [33.15983607, 53.84016393],
        [28.96721311, 47.03278689]]))


Now Compare P-Value with Alpha i.e 0.05.

- We notice that P-Value < Alpha, Therefore **REJECT** H0.

            (YES, there a relationship between day & smokers)

<br>
_____________________________________________________________________________________________________________________________